test/Test/Database/LSMTree/Internal/Serialise/Class.hs

```diff
@@ -0,0 +1,88 @@
+
+tests :: TestTree
+tests = testGroup "Test.Database.LSMTree.Internal.Serialise.Class"
+    [ testGroup "Word64"          (allProperties @Word64)
+    , testGroup "ByteString"      (allProperties @ByteString)
+    , testGroup "LazyByteString"  (allProperties @LazyByteString)
+    , testGroup "ShortByteString" (allProperties @ShortByteString)
+    , testGroup "ByteArray"       (valueProperties @ByteArray)
+    ]
+
+allProperties :: forall a. (Ord a, Show a, Arbitrary a, SerialiseKey a, SerialiseValue a) => [TestTree]
+allProperties = keyProperties @a <> valueProperties @a
+
+keyProperties :: forall a. (Ord a, Show a, Arbitrary a, SerialiseKey a) => [TestTree]
+keyProperties =
+    [ testProperty "prop_roundtripSerialiseKey" $
+        prop_roundtripSerialiseKey @a
+    , testProperty "prop_orderPreservationSerialiseKey" $
+        prop_orderPreservationSerialiseKey @a
+    ]
+
+valueProperties :: forall a. (Ord a, Show a, Arbitrary a, SerialiseValue a) => [TestTree]
+valueProperties =
+    [ testProperty "prop_roundtripSerialiseValue" $
+        prop_roundtripSerialiseValue @a
+    , testProperty "prop_concatDistributesSerialiseValue" $
+        prop_concatDistributesSerialiseValue @a
+    ]
+
+prop_roundtripSerialiseKey :: forall k. (Eq k, Show k, SerialiseKey k) => k -> Property
+prop_roundtripSerialiseKey k =
+    counterexample ("serialised: " <> show (serialiseKey k)) $
+    counterexample ("deserialised: " <> show @k (deserialiseKey (serialiseKey k))) $
+      serialiseKeyIdentity k
+
+prop_orderPreservationSerialiseKey :: forall k. (Ord k, SerialiseKey k) => k -> k -> Property
+prop_orderPreservationSerialiseKey x y =
+    counterexample ("serialised: " <> show (serialiseKey x, serialiseKey y)) $
+    counterexample ("compare: " <> show (compare x y)) $
+    counterexample ("compare serialised: " <> show (compare (serialiseKey x) (serialiseKey y))) $
+      serialiseKeyPreservesOrdering x y
+
+prop_roundtripSerialiseValue :: forall v. (Eq v, Show v, SerialiseValue v) => v -> Property
+prop_roundtripSerialiseValue v =
+    counterexample ("serialised: " <> show (serialiseValue v)) $
+    counterexample ("deserialised: " <> show @v (deserialiseValue (serialiseValue v))) $
+      serialiseValueIdentity v
+
+prop_concatDistributesSerialiseValue :: forall v. (Ord v, Show v, SerialiseValue v) => v -> Property
+prop_concatDistributesSerialiseValue v =
+    forAllShrink (genChunks bytes) shrinkChunks $ (. map (RB.pack)) $ \chs ->
+      counterexample ("from chunks: " <> show (deserialiseValueN @v chs)) $
+      counterexample ("from whole: " <> show (deserialiseValue @v (mconcat chs))) $
+        serialiseValueConcatDistributes (Proxy @v) chs
+  where
+    bytes = RB.unpack (serialiseValue v)
+
+-- | Randomly splits the input list into non-empty chunks.
+genChunks :: [a] -> Gen [[a]]
+genChunks [] = pure []
+genChunks xs = do
+    n <- chooseInt (1, length xs)
+    let (pre, post) = splitAt n xs
+    (pre :) <$> genChunks post
```

Comment:
I believe this would generate chunks that are skewed: i.e., larger chunks at the front, smaller chunks at the back (something like halving the length of the input list each time).

One could generate a number of uniformly random numbers, sort that, and use those numbers as indexes where to split the list (maybe that's a fairer splitting approach?). You could change this if you want to, but otherwise I'm fine with it as is.

Reply:
It certainly could be improved. One IMO nice aspect of the current implementation is that the expected number of chunks only grows logarithmically with the length of the list, but one could of course still have that without skew.

I'll leave it, it should work alright here.